library/psa_crypto.c

```diff
@@ -1260,6 +1260,148 @@ static psa_status_t psa_validate_optional_attributes(
+    if (alg == PSA_ALG_NONE) {
+        return PSA_SUCCESS;
+    }
+
+    /* It is not mandatory to set the key bits field, so zero is valid*/
+    if (key_bits_attr == 0) {
+        return PSA_SUCCESS;
+    }
+
+    /* Check if the size matches the curve family */
+    switch (PSA_KEY_TYPE_ECC_GET_FAMILY(type)) {
+        case PSA_ECC_FAMILY_SECP_K1:
+            if (key_bits_attr != 192 && key_bits_attr != 225 && key_bits_attr != 256) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECP_R1:
+            if (key_bits_attr != 192 && key_bits_attr != 224 && key_bits_attr != 256 &&
+                key_bits_attr != 384 && key_bits_attr != 521) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECT_K1:
+            if (key_bits_attr != 233 && key_bits_attr != 239 && key_bits_attr != 283 &&
+                key_bits_attr != 409 && key_bits_attr != 571) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECT_R1:
+            if (key_bits_attr != 233 && key_bits_attr != 283 && key_bits_attr != 409 &&
+                key_bits_attr != 571) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_BRAINPOOL_P_R1:
+            if (key_bits_attr != 192 && key_bits_attr != 224 && key_bits_attr != 256 &&
+                key_bits_attr != 320 && key_bits_attr != 384 && key_bits_attr != 512) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_TWISTED_EDWARDS:
+        case PSA_ECC_FAMILY_MONTGOMERY:
+            if (key_bits_attr != 255 && key_bits_attr != 448) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        default:
+            return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    return PSA_SUCCESS;
+}
+
+static psa_status_t psa_validate_ecc_key_data_length(const psa_key_attributes_t *attributes,
+                                                     size_t data_length)
+{
+    psa_algorithm_t alg = psa_get_key_algorithm(attributes);
+    psa_key_type_t type = psa_get_key_type(attributes);
+    size_t key_bits_attr = psa_get_key_bits(attributes);
+
+
+    /* Without  algorithm we can't evaluate more fields */
+    if (alg == PSA_ALG_NONE) {
+        return PSA_SUCCESS;
+    }
+
+        /* It is not mandatory to set the key bits field, so zero is valid */
+    if (key_bits_attr == 0) {
+        return PSA_SUCCESS;
+    }
+
+    /* Check if the size matches the curve family */
+    switch (PSA_KEY_TYPE_ECC_GET_FAMILY(type)) {
+        case PSA_ECC_FAMILY_SECP_R1:
+            /* secpr1p521 can be encoded in 65(first byte = 0) or 66 bytes therefore checking for
+             * 65 bytes is enough
+             */
+            if ((data_length < 24 && key_bits_attr == 192) ||
```

Comment:
This function is always called after we already checked the that the key_attributes is valid. 
So does it not make sense here to just write

if (data_length < PSA_BITS_TO_BYTES(key_bits_attr) 

Instead of checking for every case individually? 

Also, don't we need to return buffer_too_small or something here?

Reply:
buffer to small  is for output buffers only. invalid argument is correct for wrong size buffer on e.g. import_key

Reply:
you're probably right about the not having to check every the case

Reply:
True it came from the `secpr1p521` exception that `PSA_BITS_TO_BYTES(521)` will return 66 but 65 is also valid... due to the curve being wired. 

But can change this to have it only for that exception

Reply:
Bye looking at: https://arm-software.github.io/psa-api/crypto/1.1/api/keys/management.html#c.psa_import_key
`PSA_ERROR_BUFFER_TOO_SMALL`  is not even listed that's and in that case it is an input argument that's why went for `PSA_ERROR_INVALID_ARGUMENT`